Add function to create a random vector within a given magnitude range

hrueter opened the pull request:
## Proposed changes

Given a magnitude range return a random vector with a random magnitude in this range.
The application for this is in particular to generate random vectors that are guaranteed to be non-zero.

This code could be used in RandomUnitNormal.cpp to reduce code as they have some some partial overlap in functionality.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

Commit:
Add function to create a random vector within a given magnitude range

wthrowe commented:
Is there a reason to reimplement the angular generation rather than just multiplying a random unit normal by a random magnitude?

hrueter commented:
Good point actually.
I think I wanted to avoid generating the unnecessary flat metric first. But I think I will change it to your suggestion.

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+/// \endcond
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Construct non-zero vector by sampling spherical coordinates
```
Clarify that this only works for spatial vectors

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+///
+/// The Euclidian magnitude is sampled uniformly in the interval [min_magnitude,
+/// max_magnitude].
+/// The angle \f$\phi\f$ is sampled unifrormly in \f$[0,2 \pi]\f$.
```
Typo "uniformly"

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+    const double min_magnitude, const double max_magnitude) noexcept {
+  // check dimension
+  static_assert(Dim < 4,
+                "Function is only implemented for vectors up to rank 4.");
```
Clarify which function in the error message

hrueter replied:
I think adding the function name is not necessary, since the compiler will print it directly after the assert failed.

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+  std::piecewise_constant_distribution<> dist_signedr(
+      signedr_intervals.begin(), signedr_intervals.end(), weights.begin());
+  std::uniform_real_distribution<> dist_phi(0.0 * is_dim_greater1,
+                                            M_PI * is_dim_greater1);
```
Help text says you can go up to 2*pi, but this looks like it only draws from [0, pi]

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+Vector<DataType, Dim, Ul, Fr> make_random_vector_in_magnitude_range(
+    const gsl::not_null<std::mt19937*> nn_generator, const T used_for_size,
+    const double min_magnitude, const double max_magnitude) noexcept {
+  // check dimension
```
Why are you allowing min and max magnitudes to be nonnegative?

geoffrey4444 commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,80 @@
+
+  // generate distributions
+  const std::vector<double> signedr_intervals{-max_magnitude, -min_magnitude,
+                                              min_magnitude, max_magnitude};
```
Why does it make sense to allow negative magnitudes?

geoffrey4444 commented:
Clang-tidy failures...rebase?

hrueter commented:
I reworked the function according to @wthrowe suggestion. Since this almost changed everything I squashed immediately.
In passing I also addressed @geoffrey4444 s comments.

wthrowe commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,106 @@
+/// \cond
+namespace gsl {
+template <typename T>
+class not_null;
```
Remove this.  (You are already including the header.)

hrueter replied:
done.

wthrowe commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,106 @@
+class not_null;
+}  // namespace gsl
+
+namespace {
```
No anonymous namespaces in header files.  Use a `_detail` namespace or just inline it, since it looks like you only use it once.

hrueter replied:
done.

wthrowe commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,106 @@
+/// \ingroup TestingFrameworkGroup
+/// \brief Construct a spatial vector in a given magnitude range
+///
+/// The magnitude is compute with respect to the given metric, where the metric
```
"is computed"

hrueter replied:
fixed

wthrowe commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,106 @@
+    ERROR("min_magnitude < 0. Magnitude must be non-negative");
+  }
+  if (max_magnitude < 0) {
+    ERROR("max_magnitude < 0. Magnitude must be non-negative");
```
Should this require `min_magnitude <= max_magnitude`?

hrueter replied:
We could, but the code would work just fine if the borders were inverted, so I didn't want to restrict the user here.

hrueter replied:
I changed the tests to check that also `min_magnitude <= max_magnitude` gives correct results.

wthrowe commented on tests/Unit/Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp:
```diff
@@ -0,0 +1,106 @@
+/// \ingroup TestingFrameworkGroup
+/// \brief Construct a spatial vector in a given magnitude range
+///
+/// The magnitude is compute with respect to the flat space Euclidian metric.
```
"is computed"

hrueter replied:
fixed

wthrowe commented on tests/Unit/Helpers/Tests/Test_MakeRandomVectorInMagnitudeRange.cpp:
```diff
@@ -0,0 +1,149 @@
+#include "Helpers/DataStructures/MakeRandomVectorInMagnitudeRange.hpp"
+
+template <typename DataType, size_t Dim, typename Fr = Frame::Inertial>
+tnsr::ii<DataType, Dim, Fr> random_spatial_metric(
```
This function already exists in `Helpers/PointwiseFunctions/GeneralRelativity/TestHelpers.hpp`.

hrueter replied:
I replaced it with the helper function. I also replaced it in all other tests that reimplement `random_spatial_metric`, so that nobody will follow my example again.

I also had to adapt the helpers function, so that I can create the metric in any frame.

wthrowe commented on tests/Unit/Helpers/Tests/Test_MakeRandomVectorInMagnitudeRange.cpp:
```diff
@@ -0,0 +1,149 @@
+// Compute Magnitude
+// These functions are a workaround using Requires to
+// figure out if the inverse metric is needed
+template <typename DataType, size_t Dim, UpLo Ul, typename Fr,
```
Would it work to define overloads taking `tnsr::I` and `tnsr::i` instead of using the `Requires` machinery?

hrueter replied:
I don't think so. The function has no argument that would have type `tnsr::I` or `tnsr:i`.

wthrowe replied:
Isn't `Tensor<DataType, Symmetry<1>, index_list<SpatialIndex<Dim, UpLo::Up, Fr>>>` the same as `tnsr::I<DataType, Dim, Fr>`?

hrueter replied:
Sorry, you are right. I did not realize that the comment was about the test file. I changed it as suggested.

wthrowe commented on tests/Unit/Helpers/Tests/Test_MakeRandomVectorInMagnitudeRange.cpp:
```diff
@@ -0,0 +1,149 @@
+  return magnitude(vector, inverse_metric);
+}
+
+// Test copied from Test_RandomUnitNormal
```
I don't see anything in `Test_RandomUnitNormal.cpp` that looks like this.

hrueter replied:
I removed the comment. Seems the implementation started deviating to much.

wthrowe commented on tests/Unit/Helpers/Tests/Test_MakeRandomVectorInMagnitudeRange.cpp:
```diff
@@ -0,0 +1,149 @@
+      make_random_vector_in_magnitude_range<DataType, Dim, Ul, Fr>(
+          generator, metric, r1, r2);
+
+  Scalar<DataType> vector2_mag = magnitude_auto_invert(vector2, metric);
```
const

wthrowe commented on tests/Unit/Helpers/Tests/Test_MakeRandomVectorInMagnitudeRange.cpp:
```diff
@@ -0,0 +1,149 @@
+  std::uniform_real_distribution<> distribution1(0.0, 99.0);
+  const double r1 = distribution1(*generator);
+  std::uniform_real_distribution<> distribution2(r1, 100.0);
+  const double r2 = distribution2(*generator);
```
You should also check any edge cases you care about, such as `r1 = 0`, `r1 = r2`, or `r1 = r2 = 0`.  (I don't know which of those are considered valid.)

hrueter replied:
All of them are valid. I added tests for the two last edge cases you suggested.

hrueter replied:
I removed testing the `counter` variable, because in the edge case of `r1 = r2` the duplicate entries would no be inserted into the `values` list.
I think it wasn't really testing anything relevant anyways.

wthrowe commented:
clang-tidy has a false positive and a C++-17 cleanup for you to deal with.  Otherwise this looks good.  Go ahead and squash.

hrueter commented:
Clang-tidy has been pacified and the fixup has been squashed.

nilsdeppe commented:
@geoffrey4444 can you please check if you're happy with this now?
